lib/Dialect/Pipeline/Transforms/ScheduleLinearPipeline.cpp

```diff
@@ -0,0 +1,202 @@
+  if (!opLibAttr) {
+    pipeline.emitError("missing 'operator_lib' attribute");
+    return signalPassFailure();
+  }
+  auto opLib = dyn_cast_or_null<ssp::OperatorLibraryOp>(
+      SymbolTable::lookupNearestSymbolFrom(pipeline->getParentOp(), opLibAttr));
+  if (!opLib) {
+    pipeline.emitError("operator library '") << opLibAttr << "' not found";
+    return signalPassFailure();
+  }
+
+  auto stageOpIt = pipeline.getOps<PipelineStageOp>();
+  auto stageRegOpIt = pipeline.getOps<PipelineStageRegisterOp>();
+
+  if ((std::distance(stageOpIt.begin(), stageOpIt.end()) != 0) ||
+      (std::distance(stageRegOpIt.begin(), stageRegOpIt.end()) != 0)) {
+    pipeline.emitError("Pipeline cannot have any stages or stage registers.");
+    return signalPassFailure();
+  }
+
+  // Load operator info from attribute.
+  auto problem = Problem::get(pipeline);
+
+  DenseMap<SymbolRefAttr, Problem::OperatorType> operatorTypes;
+  SmallDenseMap<StringAttr, unsigned> oprIds;
+
+  // Set operation operator types.
+  auto returnOp =
+      cast<pipeline::ReturnOp>(pipeline.getBodyBlock()->getTerminator());
+  for (auto &op : pipeline.getOps()) {
+    // Skip if is a known non-functional operator
+    if (ignoreOp(&op))
+      continue;
+
+    Problem::OperatorType operatorType;
+    bool isReturnOp = &op == returnOp.getOperation();
+    if (isReturnOp) {
+      // Construct an operator type for the return op (not an externally defined
+      // operator type since it is intrinsic to this pass).
+      operatorType = problem.getOrInsertOperatorType("return");
+      problem.setLatency(operatorType, 0);
+    } else {
+      // Lookup operator info.
+      auto operatorTypeAttr = op.getAttrOfType<SymbolRefAttr>("operator_type");
+      if (!operatorTypeAttr) {
+        op.emitError() << "Expected 'operator_type' attribute on operation.";
+        return signalPassFailure();
+      }
+
+      auto operatorTypeIt = operatorTypes.find(operatorTypeAttr);
+      if (operatorTypeIt == operatorTypes.end()) {
+        // First time seeing operator type - load it into the problem.
+        auto opTypeOp =
+            opLib.lookupSymbol<ssp::OperatorTypeOp>(operatorTypeAttr);
+        if (!opTypeOp) {
+          op.emitError() << "Operator type '" << operatorTypeAttr
+                         << "' not found in operator library.";
+          return signalPassFailure();
+        }
+
+        auto insertRes = operatorTypes.try_emplace(
+            operatorTypeAttr, ssp::loadOperatorType<Problem, ssp::LatencyAttr>(
+                                  problem, opTypeOp, oprIds));
+        operatorTypeIt = insertRes.first;
+      }
+      operatorType = operatorTypeIt->second;
+    }
+
+    problem.insertOperation(&op);
+    problem.setLinkedOperatorType(&op, operatorType);
+
+    // We want the return op to be a sink node for all operations in the
+    // scheduling graph.
+    // To do this, we must ensure a def-use constrain exists (transitively)
```

Comment:
Better not to say "def-use constraint" here. The current terminology in the infra classifies the dependences as either auxiliary or def-use, so this could be confusing.

How about:

We want the return op to be a sink node for the dependence graph, i.e. it should (transitively) depend on every other op. This is done by inserting auxiliary dependences from ops without users, complementing the implicit dependences from the return op's operands.